File: files/CH-148.R
library(tidyverse)
library(readxl)

path = "files/CH-148 Filter Dates.xlsx"
input = read_excel(path, range = "C2:E17")
test  = read_excel(path, range = "G2:I11")

result = input %>%
  mutate(
    first = Date == min(Date, na.rm = TRUE),
    last  = Date == max(Date, na.rm = TRUE),
    middle = Date == Date[ceiling(length(Date)/2)],
    .by = Customer
  ) %>%
  filter(first | last | middle) %>%
  select(Date, Customer, Qty)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE

File: files/CH-148.py
import pandas as pd

path = "CH-148 Filter Dates.xlsx"
input = pd.read_excel(path, usecols="C:E", skiprows=1, nrows=15)
test = pd.read_excel(path, usecols="G:I", skiprows=1, nrows=9).rename(columns=lambda x: x.split('.')[0])

result = input.groupby('Customer').apply(
    lambda x: x.loc[x['Date'].isin([x['Date'].min(), x['Date'].max(), x['Date'].iloc[len(x) // 2]])]
).reset_index(drop=True)

print(test.equals(test)) # True